projection/src/test/scala/akka/projection/r2dbc/EventSourcedPubSubSpec.scala

```diff
@@ -0,0 +1,240 @@
+      (1 to 10).foreach { n =>
+        val p = n % numberOfEntities
+        entities(p) ! Persister.Persist(mkEvent(n))
+      }
+
+      val projectionName = UUID.randomUUID().toString
+      val processedProbe = createTestProbe[Processed]()
+      val slowEvents = Set(mkEvent(31), mkEvent(32), mkEvent(33))
+      val whenDone: EventEnvelope[String] => Future[Done] = { env =>
+        if (slowEvents.contains(env.event))
+          akka.pattern.after(500.millis)(Future.successful(Done))
+        else
+          Future.successful(Done)
+      }
+      val projections = startProjections(entityType, projectionName, nrOfProjections = 4, processedProbe.ref, whenDone)
+
+      // give them some time to start before writing more events, but it should work anyway
+      Thread.sleep(500)
+
+      (11 to 20).foreach { n =>
+        val p = n % numberOfEntities
+        entities(p) ! Persister.Persist(mkEvent(n))
+      }
+
+      var processed = Vector.empty[Processed]
+      processed :++= expectProcessed(processedProbe, 1, 20)
+
+      (21 to 30).foreach { n =>
+        val p = n % numberOfEntities
+        entities(p) ! Persister.Persist(mkEvent(n))
+      }
+      processed :++= expectProcessed(processedProbe, 21, 30)
+
+      // Processing of 31 is slow in the handler, see whenDone above.
+      // This will overflow the buffer for the subscribers, simulating lost messages,
+      // but they should be picked by the queries.
+      (31 to numberOfEvents - 10).foreach { n =>
+        val p = n % numberOfEntities
+        entities(p) ! Persister.Persist(mkEvent(n))
+      }
+      processed :++= expectProcessed(processedProbe, 31, numberOfEvents - 10)
+
+      (numberOfEvents - 10 + 1 to numberOfEvents).foreach { n =>
+        val p = n % numberOfEntities
+        entities(p) ! Persister.Persist(mkEvent(n))
+      }
+
+      processed :++= expectProcessed(processedProbe, numberOfEvents - 10 + 1, numberOfEvents)
+
+      val byPid = processed.groupBy(_.envelope.persistenceId)
+      byPid.foreach { case (pid, processedByPid) =>
+        // all events of a pid must be processed by the same projection instance
+        processedByPid.map(_.projectionId).toSet.size shouldBe 1
+        // processed events in right order
+        processedByPid.map(_.envelope.sequenceNr) shouldBe (1 to processedByPid.size).toVector
+
+        val viaPubSub =
+          processedByPid.filter(p =>
+            p.envelope.offset.asInstanceOf[TimestampOffset].timestamp == p.envelope.offset
+              .asInstanceOf[TimestampOffset]
+              .readTimestamp)
+        log.info("via pub-sub {}: {}", pid, viaPubSub.map(_.envelope.sequenceNr).mkString(", "))
+      }
+
+      val countViaPubSub = processed.count(p =>
+        p.envelope.offset.asInstanceOf[TimestampOffset].timestamp == p.envelope.offset
+          .asInstanceOf[TimestampOffset]
+          .readTimestamp)
+      log.info("Total via pub-sub: {}", countViaPubSub)
```

Comment:
The log output looks like this. Note that the test is simulating dropped messages by using a small buffer size.

```
via pub-sub TestEntity-1|p7: 2, 3
via pub-sub TestEntity-1|p17: 1, 2, 10
via pub-sub TestEntity-1|p6: 2, 3, 4
via pub-sub TestEntity-1|p10: 2, 3
via pub-sub TestEntity-1|p13: 1, 2, 10
via pub-sub TestEntity-1|p4: 2, 3, 4
via pub-sub TestEntity-1|p9: 2, 3, 4, 5, 6, 7
via pub-sub TestEntity-1|p18: 1, 2, 3, 10
via pub-sub TestEntity-1|p0: 1, 2, 3, 4, 5, 6, 7, 8, 10
via pub-sub TestEntity-1|p5: 2, 3, 4, 5
via pub-sub TestEntity-1|p15: 1, 2, 10
via pub-sub TestEntity-1|p1: 2, 3, 4
via pub-sub TestEntity-1|p2: 2, 3, 4
via pub-sub TestEntity-1|p11: 1, 2, 3, 10
via pub-sub TestEntity-1|p16: 1, 2, 3, 10
via pub-sub TestEntity-1|p14: 1, 2, 10
via pub-sub TestEntity-1|p3: 2, 3, 4
via pub-sub TestEntity-1|p19: 1, 2, 10
via pub-sub TestEntity-1|p12: 1, 2, 10
via pub-sub TestEntity-1|p8: 2, 3
Total via pub-sub: 70
```

With default (large) buffer size:

```
via pub-sub TestEntity-1|p7: 2, 3, 4, 5, 6, 7, 8, 9, 10
via pub-sub TestEntity-1|p17: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
via pub-sub TestEntity-1|p6: 2, 3, 4, 5, 6, 7, 8, 9, 10
via pub-sub TestEntity-1|p10: 2, 3, 4, 5, 6, 7, 8, 9, 10
via pub-sub TestEntity-1|p13: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
via pub-sub TestEntity-1|p4: 2, 3, 4, 5, 6, 7, 8, 9, 10
via pub-sub TestEntity-1|p9: 2, 3, 4, 5, 6, 7, 8, 9, 10
via pub-sub TestEntity-1|p18: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
via pub-sub TestEntity-1|p0: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
via pub-sub TestEntity-1|p5: 2, 3, 4, 5, 6, 7, 8, 9, 10
via pub-sub TestEntity-1|p15: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
via pub-sub TestEntity-1|p1: 2, 3, 4, 5, 6, 7, 8, 9, 10
via pub-sub TestEntity-1|p2: 2, 3, 4, 5, 6, 7, 8, 9, 10
via pub-sub TestEntity-1|p11: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
via pub-sub TestEntity-1|p16: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
via pub-sub TestEntity-1|p14: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
via pub-sub TestEntity-1|p3: 2, 3, 4, 5, 6, 7, 8, 9, 10
via pub-sub TestEntity-1|p19: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
via pub-sub TestEntity-1|p12: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
via pub-sub TestEntity-1|p8: 2, 3, 4, 5, 6, 7, 8, 9, 10
Total via pub-sub: 190
```